### PR TITLE
Feature/test suite

### PR DESCRIPTION
**Pull request by CompuCell3D**
Implemented a regression test suite that checks energy computations, lattice monitors (e.g. volume, surface, center of mass etc...) and also indirectly PIF Steppable. This is a simple infrastructure to easily add and maintain regression tests for any plugins

### CODE REVIEW

File: CompuCell3D/core/CompuCell3D/Potts3D/Potts3D.cpp [hunk 1]
```diff
@@ -735,377 +736,398 @@ unsigned int Potts3D::metropolisFast(const unsigned int steps, const double temp
...
+                            potts_test_data.connectivity_energy = connectivityConstraint->changeEnergy(changePixel, cell, changePixelCell);
+                            if (potts_test_data.connectivity_energy) {
+                                potts_test_data.pixelCopyAccepted = false;
+                            }
+                        }
 
-				if (numberOfThreads == 1) {
-					energyCalculator->set_aceptance_probability(prob);
-				}
+                        energyCalculator->log_output(potts_test_data);
```
Comment (tjsego) [inline CompuCell3D/core/CompuCell3D/Potts3D/Potts3D.cpp:937]: A useful but unnecessary feature that's probably beyond the scope of this PR might be to add support for testing using all Metropolis implementations.

File: CompuCell3D/core/CompuCell3D/plugins/FocalPointPlasticity/FocalPointPlasticityPlugin.cpp [hunk 2]
```diff
@@ -452,7 +452,11 @@ double FocalPointPlasticityPlugin::tryAddingNewJunction(const Point3D &pt,const
 	// Randomly shuffle neighbor selection
 	std::vector<unsigned int> randNeighborIndex = std::vector<unsigned int>(maxNeighborIndexLocal + 1, 0);
 	for (unsigned int nIdx = 0; nIdx <= maxNeighborIndexLocal; ++nIdx) randNeighborIndex[nIdx] = nIdx;
-	std::random_shuffle(randNeighborIndex.begin(), randNeighborIndex.end());
+
+    // in a regular run we do randomly select a neighbor with which we make a new link. In test run or test data generate run we do not perform this randomization
+    if (!potts->get_test_output_generate_flag() && !potts->get_test_run_flag()) {
```
Comment (tjsego) [inline CompuCell3D/core/CompuCell3D/plugins/FocalPointPlasticity/FocalPointPlasticityPlugin.cpp:457]: If we were to record the seed and reuse it during testing, wouldn't this be unnecessary?

**Review by tjsego — CHANGES_REQUESTED** [review]
Overall this looks great. Just missing one file, and some minor feedback to discuss.

File: CompuCell3D/core/CompuCell3D/plugins/FocalPointPlasticity/FocalPointPlasticityPlugin.cpp (hunk 2 above)
Comment (CompuCell3D) [reply CompuCell3D/core/CompuCell3D/plugins/FocalPointPlasticity/FocalPointPlasticityPlugin.cpp:457]: Possibly. We can make separate PR testing this approach. For now I decided to play it safely. I would suggest making  an issue to track it

File: CompuCell3D/core/CompuCell3D/Potts3D/Potts3D.cpp (hunk 1 above)
Comment (CompuCell3D) [reply CompuCell3D/core/CompuCell3D/Potts3D/Potts3D.cpp:937]: Good point, I was thinking about it too but current tests focus on energy functions and (indirectly) lattice monitors. As you can see the test data generation mode records pixel copy neighbors, energy function calculations and then those are "replayed" in a separate function that is independent from various metropolis functions. So one thing we could do (separate PR ) is to add test data generation to metropolisBoundaryWalker and maybe seriously consider removing metrolpolisList function (if it still exists...)

File: CompuCell3D/core/CompuCell3D/Potts3D/Potts3D.cpp [hunk 3]
```diff
@@ -1114,226 +1136,291 @@ unsigned int Potts3D::metropolisBoundaryWalker(const unsigned int steps, const d
...
+            Point3D changePixelNeighbor = potts_test_data_local.changePixelNeighbor;
+
+            CellG *cell = cellFieldG->getQuick(changePixelNeighbor);
+            CellG* changePixelCell = cellFieldG->getQuick(changePixel);
+            unsigned int currentWorkNodeNumber = 0;
+            flipNeighborVec[currentWorkNodeNumber] = changePixelNeighbor;
+                        
+            double change = energyCalculator->changeEnergy(changePixel, cell, changePixelCell, pixel_copy_id); 
+            
+            pixel_copy_id += 1;
```
[the remark this answers is not in the document]
Comment (CompuCell3D) [reply CompuCell3D/core/CompuCell3D/Potts3D/Potts3D.cpp:1338]: fixed

File: CompuCell3D/core/CompuCell3D/Potts3D/Potts3D.cpp [hunk 4]
```diff
@@ -1114,226 +1136,291 @@ unsigned int Potts3D::metropolisBoundaryWalker(const unsigned int steps, const d
...
+    return flips;
+
+}
 
-	} //pragma omp parallel
+unsigned int Potts3D::metropolisTestRun(const unsigned int steps, const double temp) {
+
+    flipNeighborVec.assign(pUtils->getMaxNumberOfWorkNodesPotts(), Point3D());
+    
+    std::string simulation_test_data = simulation_input_dir + "/" + "potts_data_output.csv";
```
[the remark this answers is not in the document]
Comment (CompuCell3D) [reply CompuCell3D/core/CompuCell3D/Potts3D/Potts3D.cpp:1318]: we could . fixed

File: CompuCell3D/core/CompuCell3D/Potts3D/EnergyFunctionCalculatorTestDataGeneration.h [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+       virtual double changeEnergy(Point3D &pt, const CellG *newCell, const CellG *oldCell, const unsigned int _flipAttempt);
+       void setPotts(Potts3D * _potts) { potts = _potts; }
+       void setSimulator(Simulator * _sim) { sim = _sim; }
+
+       virtual void  get_current_mcs_accepted_mask_npy_array(int * intvec, int n) {}
+       virtual void  get_current_mcs_prob_npy_array(double * doublevec, int n) {}
+       virtual void  get_current_mcs_flip_attempt_points_npy_array(short * shortvec, int n) {}
+
+       virtual void setLastFlipAccepted(bool _accept) { lastFlipAccepted = _accept; }
+       virtual void set_aceptance_probability(double _prob) { }
```
[the remark this answers is not in the document]
Comment (CompuCell3D) [reply CompuCell3D/core/CompuCell3D/Potts3D/EnergyFunctionCalculatorTestDataGeneration.h:30]: Fixed

[the remark this answers is not in the document]
Comment (CompuCell3D) [reply CompuCell3D/core/CompuCell3D/Potts3D/EnergyFunctionCalculatorTestDataGeneration.h:30]: @tjsego which file is missing?

[the remark this answers is not in the document]
Comment (tjsego) [reply CompuCell3D/core/CompuCell3D/Potts3D/EnergyFunctionCalculatorTestDataGeneration.h]: PottsTestData.cpp

[the remark this answers is not in the document]
Comment (CompuCell3D) [reply CompuCell3D/core/CompuCell3D/Potts3D/EnergyFunctionCalculatorTestDataGeneration.h]: Interesting :). Fixed, indeed it was missing :)

File: CompuCell3D/core/CompuCell3D/Potts3D/CMakeLists.txt [hunk 6]
```diff
@@ -1,14 +1,17 @@
...
   DefaultAcceptanceFunction.cpp
   CustomAcceptanceFunction.cpp
   EnergyFunctionCalculator.cpp
   EnergyFunctionCalculatorStatistics.cpp
+  EnergyFunctionCalculatorTestDataGeneration.cpp
   TypeTransition.cpp
   StandardFluctuationAmplitudeFunctions.cpp
-  Potts3D.cpp)
+  Potts3D.cpp
+  PottsTestData.cpp
```
[the remark this answers is not in the document]
Comment (CompuCell3D) [reply CompuCell3D/core/CompuCell3D/Potts3D/CMakeLists.txt:13]: Fixed

**Review by CompuCell3D — COMMENTED** [review]
This looks ready for merge

File: CompuCell3D/core/CompuCell3D/Potts3D/Potts3D.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (CompuCell3D) [reply CompuCell3D/core/CompuCell3D/Potts3D/Potts3D.cpp]: fixed now

File: CompuCell3D/core/CompuCell3D/Potts3D/CMakeLists.txt (hunk 6 above)
[the remark this answers is not in the document]
Comment (tjsego) [reply CompuCell3D/core/CompuCell3D/Potts3D/CMakeLists.txt:13]: Looks good.

File: CompuCell3D/core/CompuCell3D/plugins/FocalPointPlasticity/FocalPointPlasticityPlugin.cpp (hunk 2 above)
Comment (tjsego) [reply CompuCell3D/core/CompuCell3D/plugins/FocalPointPlasticity/FocalPointPlasticityPlugin.cpp:457]: Works for me.

**Review by tjsego — APPROVED** [review]

**Review by JulianoGianlupi — APPROVED** [review]

**Pull request merged**